config/internal/configsource/manager.go

```diff
@@ -0,0 +1,345 @@
+	close(m.watchingCh)
+
+	select {
+	case err := <-errChannel:
+		// Return the first error that reaches the channel and ignore any other error.
+		return err
+	case <-m.closeCh:
+		// This covers the case that all watchers returned ErrWatcherNotSupported.
+		return ErrSessionClosed
+	}
+}
+
+// WaitForWatcher blocks until the watchers used by WatchForUpdate are all ready.
+// This is used to ensure that the watchers are in place before proceeding.
+func (m *Manager) WaitForWatcher() {
+	<-m.watchingCh
+}
+
+// Close terminates the WatchForUpdate function and closes all Session objects used
+// in the configuration. It should be called
+func (m *Manager) Close(ctx context.Context) error {
+	var errs []error
+	for _, session := range m.sessions {
+		if err := session.Close(ctx); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	close(m.closeCh)
+	m.watchersWG.Wait()
+
+	return consumererror.Combine(errs)
+}
+
+func (m *Manager) retrieveEndAllSessions(ctx context.Context) []error {
+	var errs []error
+	for _, session := range m.sessions {
+		if err := session.RetrieveEnd(ctx); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+func (m *Manager) expandStringValues(ctx context.Context, value interface{}) (interface{}, error) {
+	switch v := value.(type) {
+	case string:
+		return m.expandConfigSources(ctx, v)
+	case []interface{}:
+		nslice := make([]interface{}, 0, len(v))
+		for _, vint := range v {
+			value, err := m.expandStringValues(ctx, vint)
+			if err != nil {
+				return nil, err
+			}
+			nslice = append(nslice, value)
+		}
+		return nslice, nil
+	case map[interface{}]interface{}:
+		nmap := make(map[interface{}]interface{}, len(v))
+		for k, vint := range v {
+			value, err := m.expandStringValues(ctx, vint)
+			if err != nil {
+				return nil, err
+			}
+			nmap[k] = value
+		}
+		return nmap, nil
+	default:
+		return v, nil
+	}
+}
+
+// expandConfigSources retrieve data from the specified config sources and injects them into
+// the configuration. The Manager tracks sessions and watcher objects as needed.
+func (m *Manager) expandConfigSources(ctx context.Context, s string) (interface{}, error) {
+	// Provisional implementation: only strings prefixed with the first character '$'
+	// are checked for config sources.
+	//
+	// TODO: Handle different ways to express config sources:
+	//
+	// 1. Concatenated with other strings (needs delimiter syntax);
+	// 2. Using spaces in its declaration;
+	// 3. Multiline;
+	//
+	if len(s) == 0 || s[0] != '$' {
+		// TODO: handle escaped $.
+		return s, nil
+	}
+
+	cfgSrcName, selector, params, err := parseCfgSrc(s[1:])
+	if err != nil {
+		return nil, err
+	}
+
+	session, ok := m.sessions[cfgSrcName]
+	if !ok {
+		// The session for this config source was not created yet.
+		cfgSrc, ok := m.configSources[cfgSrcName]
+		if !ok {
+			return nil, fmt.Errorf("config source %q not found", cfgSrcName)
+		}
+
+		session, err = cfgSrc.NewSession(ctx)
```

Comment:
For ConfigSource.NewSession docs say:
>// The code managing the returned Session object must guarantee that the object is not used
>// concurrently and that a single ConfigSource only have one Session open at any time.

Are we violating the second requirement here? If the same cfgSourceName is referenced from multiple places then we are creating multiple sessions, no?

Reply:
I'm treating the config sources as named entities. Let me try to clarify with a hypothetical example:
```yaml
config_sources:
  vault/db:
    path: secret/db
  vault/users:
    path: secret/users
```

Then on usage, there should be one session for `vault/db` and another for `vault/users` when they are referenced on the config:
```yaml
exporter:
  db_exporter:
    token: $vault/db:data.token
  some_exporter:
    username: $vault/users:data.traceruser
    password: $vault/users:data.tracerpwd
```